datadriven/src/sgpp/datadriven/algorithm/GridFactory.cpp

```diff
@@ -52,68 +51,275 @@ sgpp::base::Grid *GridFactory::createGrid(const sgpp::base::GeneralGridConfigura
-  std::vector<std::vector<size_t>> interactions;
+std::set<std::set<size_t>> sgpp::datadriven::GridFactory::getInteractions(
+    sgpp::datadriven::GeometryConfiguration config) const {
+  std::set<std::set<size_t>> interactions;
+  std::vector<std::vector<int64_t>> res = config.dim;
 
-  std::vector<int64_t> res = dim;
+  interactions = std::set<std::set<size_t>>();
+  for (StencilConfiguration stencil : config.stencils) {
+    switch (stencil.stencilType) {
+      case sgpp::datadriven::StencilType::DirectNeighbour:
+        getDirectNeighbours(interactions, config, stencil);
+        break;
+      case sgpp::datadriven::StencilType::NextHierarchicalParent:
+      case sgpp::datadriven::StencilType::AllHierarchicalParent:
+        getHierarchicalParents(interactions, config, stencil);
+        break;
+      case sgpp::datadriven::StencilType::Block:
+        getBlockInteractions(interactions, config, stencil);
+      case sgpp::datadriven::StencilType::None:
+        break;
+      default:
+        std::cout << "Stencil not found" << std::endl;
+        break;
+    }
+  }
+  return interactions;
+}
 
-  if (stencilType == sgpp::datadriven::StencilType::DN) {
-    interactions = getDirectNeighbours(res);
-  } else {
-    std::cout << "Stencil not found";
-    std::cout << std::endl;
+void sgpp::datadriven::GridFactory::getBlockInteractions(
+    std::set<std::set<size_t>>& interactions, sgpp::datadriven::GeometryConfiguration& geoConf,
+    sgpp::datadriven::StencilConfiguration& stencilConf) const {
+  auto blockLenght = stencilConf.blockLenght;
+  auto multiplicatorsPerLevel = getMultiplicatorsPerLevel(geoConf.dim);
+  auto offsetsPerLevel = getOffsetPerLevel(geoConf.dim, multiplicatorsPerLevel);
+  for (size_t i : stencilConf.applyOnLayers) {
+    auto dimensions = geoConf.dim.at(i).size();
+    auto offset = offsetsPerLevel[i];
+    auto position = std::vector<int64_t>(dimensions, 0);
+    while (!position.empty()) {
+      for (size_t j = 1; j < (size_t)2 << (blockLenght * dimensions); j++) {
+        std::set<size_t> tmp;
+        for (size_t k = 0; k < blockLenght; k++) {
+          for (size_t l = 0; l < dimensions; l++) {
+            if (j & (size_t)1 << (k * dimensions + l) &&
+                static_cast<int64_t>(position[l] + k) < geoConf.dim[i][l]) {
+              position[l] += k;
+              tmp.insert(getDataIndex(dimensions, multiplicatorsPerLevel[i], position) + offset);
+              position[l] -= k;
+            }
+          }
+        }
+        interactions.insert(tmp);
+      }
+      getNextPosition(geoConf.dim[i], position, -1);
+    }
+    interactions.insert(std::set<size_t>());
   }
+}
 
-  return interactions;
+std::vector<std::vector<size_t>> sgpp::datadriven::GridFactory::getMultiplicatorsPerLevel(
+    std::vector<std::vector<int64_t>>& imageDimensions) const {
+  auto multiplicatorsPerLevel = std::vector<std::vector<size_t>>();
```

Comment:
I would use the more concise
```suggestion
  std::vector<std::vector<size_t>> multiplicatorsPerLevel;
```
for this and similar occurrences. Using `auto` like this looks overly complex as you write out the type on the RHS anyway. (Although there's no difference in compiled code with `-O3`, I guess.) IMO `auto` is perfectly fine for complicated  types returned by functions, e.g., iterators.